modules/effects/arrow-effects/src/main/kotlin/arrow/effects/Promise.kt

```diff
@@ -0,0 +1,198 @@
+package arrow.effects
+
+import arrow.Kind
+import arrow.core.Either
+import arrow.core.Left
+import arrow.core.Right
+import arrow.effects.typeclasses.Async
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * When made, a [Promise] is empty. Until it is fulfilled, which can only happen once.
+ *
+ * A [Promise] guarantees (promises) [A] at some point in the future within the context of [F].
+ * Note that since [F] is constrained to [Async] an error can also occur.
+ */
+interface Promise<F, A> {
+
+  /**
+   * Get the promised value.
+   * Suspending the fiber running the action until the result is available.
+   *
+   * {: data-executable='true'}
+   *
+   * ```kotlin:ank
```

Comment:
all of these are missing the `main` method and the `//start..` coments that encapsulate what you want to show in the UI. See the AQL examples for the syntax